File: 2022/R/04.R
#!/usr/bin/env -S Rscript --vanilla

# solve -------------------------------------------------------------------

df <- read.table("2022/data/04", sep = ",", col.names = c("x", "y"))

parse <- function(x) {
  lapply(strsplit(x, "-"), as.integer)
}

fully_contained <- function(x, y) {
  x <- parse(x)
  y <- parse(y)
  mapply(function(a, b) {
    (a[1] <= b[1] & a[2] >= b[2]) | (a[1] >= b[1] & a[2] <= b[2])
  },
  a = x,
  b = y)

}

solution1 <- sum(with(df, fully_contained(x, y)))

overlapped <- function(x, y) {
  x <- parse(x)
  y <- parse(y)
  mapply(function(a, b) {

    FALSE |
      (a[1] >= b[1] & a[1] <= b[2]) |
      (b[1] >= a[1] & b[1] <= a[2]) |
      (a[2] >= b[1] & a[2] <= b[2]) |
      (b[2] >= a[1] & b[2] <= a[2]) |
      FALSE
  },
  a = x,
  b = y)

}

solution2 <- sum(with(df, overlapped(x, y)))

# test --------------------------------------------------------------------

solutions <- readLines("2022/solutions/04")
stopifnot(
  solution1 == solutions[1],
  solution2 == solutions[2]
)

File: 2022/python/04.py
#! /usr/bin/env python3

import numpy as np
import pandas as pd


def main() :
    df = pd.read_table("2022/data/04", dtype=str, sep=",", header=None, names=["x", "y"])
    df = df.apply(lambda x: [str_to_set(i) for i in x], axis=0)    
    solution1 = df.apply(lambda x: fully_contained(x[0], x[1]), axis=1).sum()
    solution2 = df.apply(lambda x: any_overlap(x[0], x[1]), axis=1).sum()
    return [solution1, solution2]


def str_to_set(x) :
    nums = [int(i) for i in x.split("-")]
    return set(range(nums[0], nums[1] + 1))


def fully_contained(a, b) :
    return a.intersection(b) == a or b.intersection(a) == b


def any_overlap(a, b) :
    return len(a.intersection(b)) > 0


if __name__ == "__main__" :
    results = main()
    solutions = np.loadtxt("2022/solutions/04", dtype=int)

    if np.all(results != solutions) :
        raise ValueError("solutions are incorrect")

